File: src/integrator/DynamicResolution.py
#  Copyright (c) 2015-2016,2021
#      Jakub Krajniak (jkrajniak at gmail.com)
#
#  This file is part of ESPResSo++.
#
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""
*******************************************
**espressopp.integrator.DynamicResolution**
*******************************************

*DynamicResolution* extension allows changing the *lambda* parameter of particles, namely
the so called resolution. The module can be used to perform backmapping.

The resolution is changed by

.. math::

   \lambda(t) = \lambda_0 + at


where :math:`\lambda` is the current resolution of the particles, :math:`a` is
the rate by which the resolution is changed during the simulation and :math:`\lambda_0`
is an initial resolution.

.. class:: espressopp.integrator.DynamicResolution

      The main class for DynamicResolution method.

      .. data:: resolution

         The initial resolution (:math:`\lambda_0`).

      .. data:: rate

         The rate of resolution update.

      .. data:: active

         If set to True then resolution change is active.

      .. method:: DynamicResolution(system, vs_list, rate)

         :param system: The system object.
         :type system: espressopp.System
         :param vs_list: The MD integrator.
         :type vs_list: espressopp.FixedVSList
         :param rate: The rate.
         :type output: float

      .. method:: update_weights

         Manual update the resolutions of the particles.

"""

from espressopp.esutil import cxxinit
from espressopp import pmi

from espressopp.integrator.Extension import Extension, ExtensionLocal
from _espressopp import integrator_DynamicResolution


class DynamicResolutionLocal(ExtensionLocal, integrator_DynamicResolution):
    def __init__(self, _system, _vs_list, _rate):
        'Local construction of a verlet list for AdResS'
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            cxxinit(self, integrator_DynamicResolution,
                    _system, _vs_list, _rate)

    def update_weights(self):
        if pmi.workerIsActive():
            self.cxxclass.update_weights(self)


if pmi.isController:
    class DynamicResolution(Extension, metaclass=pmi.Proxy):
        pmiproxydefs = dict(
            cls='espressopp.integrator.DynamicResolutionLocal',
            pmiproperty=['rate', 'active'],
            pmicall=['update_weights']
        )

File: src/integrator/DynamicResolution.cpp
/*
  Copyright (C) 2015-2016,2021
      Jakub Krajniak (jkrajniak at gmail.com)

  This file is part of ESPResSo++.

  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "DynamicResolution.hpp"
#include <algorithm>
#include <vector>
#include <iomanip>
#include <sstream>
#include "python.hpp"
#include "Real3D.hpp"
#include "Particle.hpp"
#include "Cell.hpp"
#include "System.hpp"
#include "storage/Storage.hpp"
#include "bc/BC.hpp"
#include "iterator/CellListAllPairsIterator.hpp"
#include "iterator/CellListIterator.hpp"

namespace espressopp
{
namespace integrator
{
using namespace espressopp::iterator;  // NOLINT

LOG4ESPP_LOGGER(DynamicResolution::theLogger, "DynamicResolution");

DynamicResolution::DynamicResolution(shared_ptr<System> _system,
                                     shared_ptr<FixedVSList> _vslist,
                                     real _rate)
    : Extension(_system), rate_(_rate), vs_list(_vslist)
{
    LOG4ESPP_INFO(theLogger, "construct DynamicResolution");
    type = Extension::Adress;
}

DynamicResolution::~DynamicResolution()
{
    LOG4ESPP_INFO(theLogger, "~DynamicResolution");
    disconnect();
}

void DynamicResolution::connect()
{
    LOG4ESPP_INFO(theLogger, "connect");
    _aftIntV = integrator->aftIntV.connect(boost::bind(&DynamicResolution::updateWeights, this),
                                           boost::signals2::at_back);
}

void DynamicResolution::disconnect()
{
    LOG4ESPP_INFO(theLogger, "disconnect");
    _aftIntV.disconnect();
}

void DynamicResolution::set_active(bool active) { active_ = active; }

void DynamicResolution::updateWeights()
{
    LOG4ESPP_INFO(theLogger, "updateWeights");
    if (!active_) return;

    System &system = getSystemRef();

    // Update weights of the particles.
    FixedVSList::GlobalTuples vs = vs_list->globalTuples;
    FixedVSList::GlobalTuples::iterator it = vs.begin();
    for (; it != vs.end(); ++it)
    {
        Particle *vp = system.storage->lookupLocalParticle(it->first);
        if (vp)
        {
            real res = vp->lambda() + rate_;
            if (res > (1.0 + rate_)) continue;
            if (res > 1.0) res = 1.0;
            vp->lambda() = res;

            // Update weights for all underlying particles.
            for (FixedVSList::tuple::iterator itp = it->second.begin(); itp != it->second.end();
                 ++itp)
            {
                Particle *at = system.storage->lookupLocalParticle(*itp);
                if (at)
                {
                    at->lambda() = res;
                }
                else
                {
                    std::cout << " AT particle (" << *itp << ") of VP " << vp->id() << "-"
                              << vp->ghost() << " not found in tuples ";
                    std::cout << " (" << vp->position() << ")" << std::endl;
                    exit(1);
                }
            }
        }
    }
}

/****************************************************
** REGISTRATION WITH PYTHON
****************************************************/
void DynamicResolution::registerPython()
{
    using namespace espressopp::python;  // NOLINT
    class_<DynamicResolution, shared_ptr<DynamicResolution>, bases<Extension> >(
        "integrator_DynamicResolution", init<shared_ptr<System>, shared_ptr<FixedVSList>, real>())
        .add_property("active", &DynamicResolution::active, &DynamicResolution::set_active)
        .add_property("rate", &DynamicResolution::rate, &DynamicResolution::set_rate)
        .def("update_weights", &DynamicResolution::updateWeights)
        .def("connect", &DynamicResolution::connect)
        .def("disconnect", &DynamicResolution::disconnect);
}
}  // end namespace integrator
}  // end namespace espressopp
